design-proposals/network-vhostuser-binding-plugin/network-vhostuser-binding-plugin.md

```diff
@@ -0,0 +1,141 @@
+
+However this assumes:
+- the `sockets` emptyDir can be used for such a purpose
+- the CNI has access to the `sockets` emptyDir of the `virt-launcher` pod and that it can bind mount it to a path available to the dataplane. This can be tricky especially with Multus 4 in thick plugin mode, where the CNI is executed by the mutlus pod. Usually the multus thick plugin daemonset defines a `/hostroot` hostPath volume mount with `mountPropagation: HostToContainer` option. A `mountPropagation: Bidirectional` option is needed for the bind mount to be propagated back to the host and to the dataplane pod.
+- the dataplane pod has the privilege to mount a hostPath
+
+Here is a diagram showing sockets sharing mecanisms between `virt-launcher` pod, `userspace CNI` and the `dataplane` pod.
+
+![kubevirt-vhostuser-shared-sockets](kubevirt-vhostuser-shared-sockets.png)
+
+Sharing the `vhostuser` sockets between `virt-launcher` pods and the dataplane pod is something to be enhanced in order to limit usage of hostPath volumes and bind mounts.
+
+## Alternative designs
+
+Some alternative designs were discussed in ![kubevirt-dev mailing-list](https://groups.google.com/g/kubevirt-dev/c/3w_WStrJfZw/m/yWSBpDAKAQAJ).
+
+### Expose a virt-launcher pod directory to binding plugin
+
+This requires to implement a new network binding plugin mechanism we could expose the content of a `virt-launcher` directory to an external plugin.
+The plugin registration in Kubervirt resource would define the target directory on the node where the directory should be exposed.
+
+This diagram explains this mechanism.
+
+![kubevirt-plugin-extension](kubevirt-plugin-extension.png)
+
+The advantages of such an approach are:
+- is generic and could also be potentially reused by other device types
+- hides the KubeVirt implementation details. Currently, you need to know where the KubeVirt sockets are located in the virt-launcher filesystem. Potentially, if we change the directory path for the sockets, this would break the CNI plugin
+- can isolate the resources dedicated to that particular plugin
+
+The drawback side is that `virt-launcher` pod would need to be run with `privileged: true` in order to do the bind mount.
+
+### Device plugin for `vhostuser sockets` resource
+
+Device plugins have the ability to add hostPath mounts to the pods when they request for managed resources.
+We could implement a vhostuser device plugin that would manage two kinds of resources:
+- dataplane: 1
+  This only resource is resquested by the userspace dataplane, and add a `/var/run/vhost_sockets` mount to the dataplane pod.
+- vhostuser sockets: n
+  This as many resources as we want to handle, is requested by the `virt-launcher` pod using vhostuser plugin. This makes the device plugin create a per pod directory like `/var/run/vhost_sockets/<launcher-id>`, and mount it into the `virt-launcher` pod.
```

Comment:
Is the pod ID know by the CNI plugin or outside KubeVirt?

Reply:
DP can push annotation to the pod with an `id` it will generate. So it's not really the virt-launcher pod id, as it does not exist at the time DP is called.
The CNI can read the pod annotation and use it to configure the vhostuser with the right path in the data plane.